sdks/go/pkg/beam/io/webapi/webapi.go

```diff
@@ -0,0 +1,209 @@
+// Licensed to the Apache Software Foundation (ASF) under one or more
+// contributor license agreements.  See the NOTICE file distributed with
+// this work for additional information regarding copyright ownership.
+// The ASF licenses this file to You under the Apache License, Version 2.0
+// (the "License"); you may not use this file except in compliance with
+// the License.  You may obtain a copy of the License at
+//
+//    http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package webapi
+
+import (
+	"bytes"
+	"context"
+	"fmt"
+	"github.com/apache/beam/sdks/v2/go/pkg/beam"
+	"github.com/apache/beam/sdks/v2/go/pkg/beam/register"
+	"reflect"
+)
+
+func init() {
+	beam.RegisterType(reflect.TypeOf((*wrappedCallerOnlyUserType)(nil)).Elem())
+	beam.RegisterType(reflect.TypeOf((*Request)(nil)).Elem())
+	beam.RegisterType(reflect.TypeOf((*Response)(nil)).Elem())
+	register.Emitter1[*Response]()
+	register.DoFn3x1[context.Context, *Request, func(*Response), error](&callerFn{})
+}
+
+// Caller is an interface to a Web API endpoint.
+type Caller interface {
+
+	// Call a Web API endpoint with a Request, yielding a Response.
+	Call(ctx context.Context, request *Request) (*Response, error)
+}
+
+// SetupTeardown interfaces methods called during a setup and teardown DoFn lifecycle.
+// Some clients to Web APIs may need to establish resources or network connections prior to
+// calling a Web API. This is a separate interface from a Caller since some Web API endpoints
+// do not require this. An internal hybrid interface is hidden for user convenience so that
+// a developer need only provide what makes sense for the API read or write task.
+type SetupTeardown interface {
+
+	// Setup is called during a DoFn setup lifecycle method.
+	// Clients that need to instantiate resources would be performed within an implementation of this method.
+	Setup(ctx context.Context) error
+
+	// Teardown is called during a DoFn teardown lifecycle method.
+	// Clients that need to destroy or close resources would be performed within an implementation of this method.
+	Teardown(ctx context.Context) error
+}
+
+// Request holds the encoded payload of a Web API request.
+// Caller implementations are responsible for decoding the Payload into the needed type required
+// to fulfill the Web API request.
+type Request struct {
+
+	// Payload is the encoded Web API request.
+	Payload []byte `beam:"payload"`
```

Comment:
Note, we can make this generic instead of forcing inefficient user side encoding & decoding, it just requires the additional registration steps due to the current SDK's limitations WRT registring types.

eg.

The closest to best experience I think we could manage is:

``` 
func init() {
   webapi.RegisterCaller(&myCaller{})
}


...
responses, failures := webapi.Call(s, &myCaller{/*config*/}, requests)
...

```

That function with the most recent Go versions (1.21 I think?) should be able to  infer the types.

Make the interface and implementation types generic, this avoids the extra wrapper types for Request and Response.

The callerFn type can then become generic on the user's caller type, the request type and the response type of the Caller interface methods, which is also pushed into the interface.